File: Solutions/C++/Medium/102.cpp
/*
    We perform Breadth-First-Search where the values of the nodes at the same depth are stored together in a vector.
*/

class Solution {
public:
    std::vector<std::vector<int>> levelOrder(TreeNode* root) {
        std::vector<vector<int>> result;
        if (!root) {
            return result;
        }
        std::deque<TreeNode*> q;
        q.push_back(root);
        while (q.size() > 0) {
            int qSize = q.size();
            std::vector<int> levelNodes;
            while (qSize > 0) {
                TreeNode* node = q.front();
                q.pop_front();
                qSize--;
                levelNodes.push_back(node->val);
                if (node->left) {
                    q.push_back(node->left);
                }
                if (node->right) {
                    q.push_back(node->right);
                }
            }
            if (levelNodes.size() > 0) {
                result.push_back(levelNodes);
            }
        }
        return result;
    }
};

File: Solutions/Python/Medium/102.py
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result: List[List[int]] = []
        if not root:
            return result
        queue: List[Optional[TreeNode]] = []
        queue.append(root)
        while queue:
            qSize: int = len(queue)
            levelNodes: List[int] = []
            while qSize:
                node = queue.pop(0)
                qSize -= 1
                levelNodes.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            result.append(levelNodes)
        return result
